### PR TITLE
Couple c14 net

### LINKED ISSUE #20: Couple C14 net into CLUBB
Coupling C14 NN trained by Alex into CLUBB to run.

Questions:

- Where to couple in the code?
- What are the inputs?
- What are the outputs?

---------

### Where?

Suggested  as part of the following loop in [`src/CLUBB_core/advance_xp2_xpyp_module.F90`](https://github.com/m2lines/clubb_ML/blob/92b8d7aeeafc1b045641b4c91806144a1c68945b/src/CLUBB_core/advance_xp2_xpyp_module.F90) as part of the `advance_xp2_xpyp` subroutine.

https://github.com/m2lines/clubb_ML/blob/92b8d7aeeafc1b045641b4c91806144a1c68945b/src/CLUBB_core/advance_xp2_xpyp_module.F90#L517-L525

This is a routine that prognoses variances and turbulence components.
Called from `src/CLUBB_core/advance_clubb_core_module.F90`.

I believe the result of the NN will replace `C14_1d(i,k)`.

## Inputs

Inputs should all be available in `advance_xp2_xpyp`.

We require, based on training notebooks:

- $u'^2$, $v'^2$, $w'^2$, each normalised by $e' = 0.5 * (u'^2 + v'^2 + w'^2)$
    - In `advance_xp2_xpyp` I believe these are `up2`, `vp2`, and `wp2`.
      u and v are `intent(in)` to `advance_xp2_xpyp` whilst w is `intent(inout)`. so need to be careful here.
- `Lup / Hscale` and `Ldown/Hscale`
    - Hscale should be a scalar constant, but I can't see where to obtain this in CLUBB
    - Lup and Ldown vary spatially, again, it's not clear to me where these come from.
      I cannot see `Lscale_up` or `Lscale_down` as appear in the mixing length code.

## Outputs

I believe the result of the NN will replace `C14_1d(i,k)`.

This is passed to `term_dp1_lhs` later in `advance_xp2_xpyp`.

---------

## Questions

- Locating of input variables.
- Is it enough to replace C14_1d here, or are there knock-on changes to be made where it is used?
    - We are not directly modifying `mixing_length` code but instead C14 feeds into this to change the calculation?

### PR DESCRIPTION
**Pull request by jatkinson1000**
Work to close #20

Builds on #24, #23, and #21

Now that FTorch is in the code and all infrastructure is set up look to specifically couple and run the C14 net.

TODO:

- [x] Bring `l_c14_ml` into `advance_xp2_xpyp` from `clubb_config_flags` which is present in `src/CLUBB_core/advance_clubb_core_module.F90` from which `advance_xp2_xpyp` is called.
- [x] Set up a separate loop for C14 that is dependant on value of the trigger
- [x] Bring input data required to appropriate location in the code
- [x] Construct input vector as appropriate
- [ ] Once working look to use batching outside of the loop

### CODE REVIEW

**Comment by jatkinson1000** [comment]
With the above code I was able to run CLUBB with the BOMEX test case but using a version of the net trained on the BOMEX data.

This runs to completion, without failure, though the next step would be to inspect the results more carefully to ensure it is doing something sensible in the physical sense and more broadly in the model.

I output the old and new C14 values being calculated for each cell at each iteration, a sample of which is shown in the dropdown below. Note that values become negative for large `k`, with the final value being positive.

<details>
<summary> Example CLUBB Output </summary>

```
iteration =      358; time =    21480.0
    2.00    4.00    6.00    8.00
Old C14:   0.3333  ML C14:   0.8343
Old C14:   0.3333  ML C14:   1.1955
Old C14:   0.3333  ML C14:   1.5988
Old C14:   0.3333  ML C14:   1.8375
Old C14:   0.3333  ML C14:   1.9465
Old C14:   0.3333  ML C14:   2.0219
Old C14:   0.3333  ML C14:   2.0746
Old C14:   0.3333  ML C14:   2.1115
Old C14:   0.3333  ML C14:   2.1365
Old C14:   0.3333  ML C14:   2.1507
Old C14:   0.3333  ML C14:   2.1530
Old C14:   0.3333  ML C14:   2.1403
Old C14:   0.3333  ML C14:   2.1088
Old C14:   0.3333  ML C14:   1.9589
Old C14:   0.3333  ML C14:   1.6371
Old C14:   0.3333  ML C14:   1.4197
Old C14:   0.3333  ML C14:   1.3243
Old C14:   0.3333  ML C14:   1.3201
...
Old C14:   0.3333  ML C14:  -0.3243
Old C14:   0.3333  ML C14:  -0.3239
Old C14:   0.3333  ML C14:  -0.3236
Old C14:   0.3333  ML C14:  -0.3233
Old C14:   0.3333  ML C14:  -0.3232
Old C14:   0.3333  ML C14:  -0.3231
Old C14:   0.3333  ML C14:  -0.3229
Old C14:   0.3333  ML C14:  -0.3227
Old C14:   0.3333  ML C14:  -0.3222
Old C14:   0.3333  ML C14:  -0.3215
Old C14:   0.3333  ML C14:  -0.3200
Old C14:   0.3333  ML C14:  -0.3173
Old C14:   0.3333  ML C14:  -0.3120
Old C14:   0.3333  ML C14:  -0.2994
Old C14:   0.3333  ML C14:  -0.2597
Old C14:   0.3333  ML C14:   0.0459
```

</details>

My suggestions would be to now meet to discuss results, evaluate if things are working as expected, and think about next steps.

---

#### Saving a net for use in CLUBB

To save the model for use in Fortran I used the following cell at the bottom of the training notebook:

```python
# Save to Torchscript
model_file = "<path/to>/CLUBBiNN/saved_models/BOMEX_model.pt"

test_in = torch.tensor([1.0, 1.0, 1.0, 1.0, 1.0], dtype=torch.float64)
test_out = best_model(test_in)
print(test_out)

# Save model to TorchScript using tracing
best_model.eval()
scripted_model = torch.jit.script(best_model)
scripted_model.save(model_file)

# Check saved model can be loaded and performs as expected
ts_model = torch.jit.load(model_file)
ts_out = ts_model(test_in)
print(ts_out)
```

This should be added more formally to the training code in CLUBBiNN to make this easy for subsequent training.
I used a notebook from when we finished refactoring but before Alex's recent commits.

**Comment by adconnolly** [comment]
Regarding the negative C14s, the BOMEX-only training notebook wasn't using a bounding strategy until recently, the "allLES" notebooks were using a clamping layer, but I also recently updated that to a sigmoid based approach to bounding so the gradients wouldn't get locked at zero during some training runs. This will be a nice opportunity to see what happens to those upper levels with the bounding approaches - I'd assume they'll go to zero and am rather curious why, from an input feature dependence point-of-view, that is the case farther aloft.

**Comment by jatkinson1000** [comment]
This was replaced by #28 due to conflicts after merging #21 and #23 (and then #27)
Worked through with @adconnolly 

The code content between the two is identical, but conflicts/additional code appears due to merge vs. rebase.
